File: web/tests/protocol.test.ts
import { describe, it, expect } from 'vitest';
import {
  parseControlMessage,
  encodeControlMessage,
  type ControlMessage,
} from '../src/lib/protocol';

function roundtrip(msg: ControlMessage): ControlMessage {
  return parseControlMessage(encodeControlMessage(msg));
}

describe('protocol', () => {
  it('encodes and decodes wrapper_hello', () => {
    const msg: ControlMessage = {
      type: 'wrapper_hello',
      api_key: 'k'.repeat(43),
      token: 't'.repeat(43),
      cols: 80,
      rows: 24,
    };
    const back = roundtrip(msg);
    expect(back.type).toBe('wrapper_hello');
    if (back.type === 'wrapper_hello') {
      expect(back.cols).toBe(80);
      expect(back.token).toBe('t'.repeat(43));
    }
  });

  it('encodes and decodes phone_hello', () => {
    const msg: ControlMessage = {
      type: 'phone_hello',
      token: 'a'.repeat(43),
      cols: 80,
      rows: 24,
    };
    const back = roundtrip(msg);
    expect(back.type).toBe('phone_hello');
    if (back.type === 'phone_hello') {
      expect(back.cols).toBe(80);
    }
  });

  it('encodes and decodes server_hello', () => {
    const msg: ControlMessage = {
      type: 'server_hello',
      session_id: 's-1',
      peer_connected: true,
    };
    const back = roundtrip(msg);
    expect(back.type).toBe('server_hello');
    if (back.type === 'server_hello') {
      expect(back.peer_connected).toBe(true);
    }
  });

  it('encodes and decodes resize', () => {
    const msg: ControlMessage = { type: 'resize', cols: 100, rows: 40 };
    const back = roundtrip(msg);
    expect(back.type).toBe('resize');
    if (back.type === 'resize') {
      expect(back.cols).toBe(100);
      expect(back.rows).toBe(40);
    }
  });

  it('encodes and decodes peer_status', () => {
    const back = roundtrip({ type: 'peer_status', connected: true });
    expect(back.type).toBe('peer_status');
  });

  it('encodes and decodes close with optional reason', () => {
    const back = roundtrip({ type: 'close', reason: 'bye' });
    expect(back.type).toBe('close');
  });

  it('rejects non-object', () => {
    expect(() => parseControlMessage('"foo"')).toThrow();
  });

  it('rejects null', () => {
    expect(() => parseControlMessage('null')).toThrow();
  });

  it('rejects object without type field', () => {
    expect(() => parseControlMessage('{"foo":"bar"}')).toThrow();
  });

  it('rejects malformed JSON', () => {
    expect(() => parseControlMessage('{not json')).toThrow();
  });

  it('parses error message variant', () => {
    const msg: ControlMessage = {
      type: 'error',
      code: 'invalid_token',
      message: 'nope',
    };
    const back = roundtrip(msg);
    expect(back.type).toBe('error');
    if (back.type === 'error') {
      expect(back.code).toBe('invalid_token');
    }
  });
});

File: crates/claude-phone-shared/tests/protocol_test.rs
use claude_phone_shared::protocol::*;
use claude_phone_shared::{ApiKey, SessionToken};

#[test]
fn wrapper_hello_roundtrip() {
    let msg = ControlMessage::WrapperHello(WrapperHello {
        api_key: ApiKey::generate(),
        token: SessionToken::generate(),
        cols: 80,
        rows: 24,
        claude_version: Some("1.2.3".into()),
    });

    let s = serde_json::to_string(&msg).unwrap();
    assert!(s.contains("wrapper_hello"));

    let back: ControlMessage = serde_json::from_str(&s).unwrap();
    match back {
        ControlMessage::WrapperHello(h) => {
            assert_eq!(h.cols, 80);
            assert_eq!(h.rows, 24);
            assert_eq!(h.claude_version.as_deref(), Some("1.2.3"));
        }
        _ => panic!("wrong variant"),
    }
}

#[test]
fn phone_hello_roundtrip() {
    let msg = ControlMessage::PhoneHello(PhoneHello {
        token: SessionToken::generate(),
        cols: 40,
        rows: 80,
        user_agent: Some("Mozilla/5.0".into()),
    });

    let s = serde_json::to_string(&msg).unwrap();
    assert!(s.contains("phone_hello"));

    let back: ControlMessage = serde_json::from_str(&s).unwrap();
    match back {
        ControlMessage::PhoneHello(h) => {
            assert_eq!(h.cols, 40);
            assert_eq!(h.rows, 80);
        }
        _ => panic!("wrong variant"),
    }
}

#[test]
fn server_hello_ok() {
    let msg = ControlMessage::ServerHello(ServerHello {
        session_id: "abc123".into(),
        peer_connected: false,
    });
    let s = serde_json::to_string(&msg).unwrap();
    assert!(s.contains("server_hello"));
    let _: ControlMessage = serde_json::from_str(&s).unwrap();
}

#[test]
fn error_message() {
    let msg = ControlMessage::Error(ErrorMessage {
        code: ErrorCode::InvalidToken,
        message: "token not found".into(),
    });
    let s = serde_json::to_string(&msg).unwrap();
    assert!(s.contains("invalid_token"));
    let _: ControlMessage = serde_json::from_str(&s).unwrap();
}

#[test]
fn resize_roundtrip() {
    let msg = ControlMessage::Resize(Resize {
        cols: 100,
        rows: 50,
    });
    let s = serde_json::to_string(&msg).unwrap();
    let back: ControlMessage = serde_json::from_str(&s).unwrap();
    match back {
        ControlMessage::Resize(r) => {
            assert_eq!(r.cols, 100);
            assert_eq!(r.rows, 50);
        }
        _ => panic!("wrong"),
    }
}

#[test]
fn peer_status() {
    let msg = ControlMessage::PeerStatus(PeerStatus { connected: true });
    let s = serde_json::to_string(&msg).unwrap();
    let _: ControlMessage = serde_json::from_str(&s).unwrap();
}

#[test]
fn close_roundtrip() {
    let msg = ControlMessage::Close(Close {
        reason: Some("user quit".into()),
    });
    let s = serde_json::to_string(&msg).unwrap();
    let _: ControlMessage = serde_json::from_str(&s).unwrap();
}

#[test]
fn unknown_type_rejected() {
    let s = r#"{"type":"bogus","x":1}"#;
    let r: Result<ControlMessage, _> = serde_json::from_str(s);
    assert!(r.is_err());
}
